From 7724f377933ac2c0fe802dcdf34e7c5a50f00422 Mon Sep 17 00:00:00 2001
From: Hoodaly <hoodaly@gmail.com>
Date: Thu, 28 Feb 2019 16:08:53 +0100
Subject: [PATCH 26/34] Skip one-shot sensors for WindowOrientationListener

The HTC One M8 exposes a custom sensor "HTC Gesture sensor" with type 27, which is now used for SENSOR_TYPE_DEVICE_ORIENTATION sensors.
It is a one-shot sensor, which cannot be registered by the WindowOrientationListener. But as the list of device orientation sensors is not empty,
no alternative is considered and therefore automatic screen rotation doesn't work.

Don't consider one-shot sensors for this listener, so gravity or acceleration sensors are used and screen rotation works.

Change-Id: I1136ad9ff9f44b171eecac579b4d9934aee83e20
---
 .../com/android/server/policy/WindowOrientationListener.java  | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/services/core/java/com/android/server/policy/WindowOrientationListener.java b/services/core/java/com/android/server/policy/WindowOrientationListener.java
index bc397737976..b81b18ee8e9 100644
--- a/services/core/java/com/android/server/policy/WindowOrientationListener.java
+++ b/services/core/java/com/android/server/policy/WindowOrientationListener.java
@@ -102,6 +102,10 @@ public abstract class WindowOrientationListener {
          *  this type.
          */
         for (Sensor s : l) {
+            // Sensors in one-shot mode are not suitable.
+            if (s.getReportingMode() == Sensor.REPORTING_MODE_ONE_SHOT) {
+                continue;
+            }
             if (s.isWakeUpSensor()) {
                 wakeUpDeviceOrientationSensor = s;
             } else {
-- 
2.17.1

